specs/beacon-chain.md

```diff
@@ -870,6 +905,81 @@ while len(crystallized_state.persistent_committee_reassignments) > 0 and crystal
+
+### Verifying a crosslink
+
+A node should sign a crosslink only if the following conditions hold. **If a node has the capability to perform the required level of verification, it should NOT follow chains on which a crosslink for which these conditions do NOT hold has been included, or a sufficient number of signatures have been included that during the next state recalculation, a crosslink will be registered.**
+
+First, the conditions must recursively apply to the crosslink referenced in `last_crosslink_hash` for the same shard (unless `last_crosslink_hash` equals zero, in which case we are at the genesis).
+
+Second, we verify the `shard_block_combined_data_root`. Let `B[0]` be the first block _after_ the last crosslink and `B[n-1]` be the block directly referenced by the `shard_block_hash`. Let `bodies[0] .... bodies[n-1]` be their bodies and `roots[0] ... roots[n-1]` the data roots. Define `compute_merkle_root` be a simple Merkle root calculating function that takes as input a list of objects, where the list's length must be an exact power of two. Let `state[0] ... state[n-1]` be the beacon chain states at those times, and `depths[0] ... depths[n-1]` be equal to `log2(next_power_of_2(calc_block_maxbytes(state[i]) // CHUNK_SIZE))` (ie. the expected depth of the i'th data tree).
+
+```python
+def get_zeroroot_at_depth(n):
+    o = b'\x00' * CHUNK_SIZE
+    for i in range(n):
+        o = hash(o + o)
+    return o
+
+def mk_combined_data_root(depths, roots):
+    default_value = get_zeroroot_at_depth(max(depths))
+    data = [default_value for _ in range(next_power_of_2(len(roots)))]
+    for i, (depth, root) in enumerate(zip(depths, roots)):
+        value = root
+        for j in range(depth, max(depths)):
+            value = hash(value, get_zeroroot_at_depth(depth + j))
+        data[i] = value
+    return compute_merkle_root(data)
+```
+
+Essentially, this outputs the root of a tree of the data roots, with the data roots all adjusted to have the same height if needed. The tree can also be viewed as a tree of all of the underlying data concatenated together, appropriately padded. Here is an equivalent definition that uses bodies instead of roots [TODO: check equivalence]:
+
+```python
+def mk_combined_data_root(depths, bodies):
+    default_value = get_zeroroot_at_depth(max(depths))
+    padded_body_length = max([CHUNK_SIZE * 2**d for d in depths])
+    data = b''
+    for body in bodies:
+        padded_body = body + bytes([0] * (padded_body_length - len(body)))
+        data += padded_body
+    data += bytes([0] * (next_power_of_2(len(data)) - len(data))
+    return compute_merkle_root([data[pos:pos+CHUNK_SIZE] for pos in range(0, len(data), CHUNK_SIZE)])
+```
+
+Verify that the `shard_block_combined_data_root` is the output of these functions.
```

Comment:
Should this verification be happening at the beacon chain level? or is this independent?